fix(gitea): allow fallbacks in case default merge style not allowed

dh-seanmurphy opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

Asked via discussion [here](https://github.com/renovatebot/renovate/discussions/37665).

As noted in #36779, the default merge style in Gitea is always set, but it's possible that it is set to a merge style which is not allowed for the repo. Currently, Renovate blocks the repo in this case.

Gitea's handing of this is less strict; if the default merge style is not an allowed merge style, Gitea suggests one of the allowed merge styles based on an ordered list.

This change mirrors the behaviour of Gitea by permitting fallbacks in case the default merge style is not an allowed merge style; the fallback is selected from an ordered list, using the same ordering as Gitea. If the default merge style is an allowed style, the behaviour is unchanged.

## Context

Please select one of the below:

- [ ] This closes an existing Issue: #
- [x] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [ ] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [x] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

I used ChatGPT to generate a first version of this and iterated on it. I obtained human feedback and then used ChatGPT to help with refactoring. Modifications to tests were performed manually.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,28 +323,35 @@ const platform: Platform = {
+      'fast-forward-only',
+    ];
+
+    const candidateMergeStyles = Array.from(
```
IMO the comment regarding why we do this belongs here, not (merely) in the PR description.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,28 +323,35 @@ const platform: Platform = {
+        selectedMergeStyle = s;
+        break;
+      }
+    }
```
Factor merge style determination into a method? (whether or not it is worth it will depend on the specific refactors above).

dh-seanmurphy replied:
Significant refactoring based on feedback. Please LMK what you think.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,28 +323,35 @@ const platform: Platform = {
+        (s === 'rebase' && repo.allow_rebase) ||
+        (s === 'rebase-merge' && repo.allow_rebase_explicit) ||
+        (s === 'squash' && repo.allow_squash_merge) ||
+        (s === 'fast-forward-only' && repo.allow_fast_forward_only_merge)
```
This will not make typescript fail if there is an additional / unhandled `PRMergeMethod`.

Options:
- Use a switch
- Use a `Record<PRMergeMethod, boolean>` to look up allowed (probably, you'll need to test).

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,28 +323,35 @@ const platform: Platform = {
+    ];
+
+    const candidateMergeStyles = Array.from(
+      new Set<PRMergeMethod>([repo.default_merge_style, ...giteaOrdering]),
```
What's the point of using a `Set` here?

- Less readable (you need to know that in JS sets have defined iteration order)
- Likely slower (to build up the set).

dh-seanmurphy replied:
I've completely refactored this to remove the set and I think it's significantly more readable/intuitive.

gzm0 commented on lib/modules/platform/gitea/index.spec.ts:
```diff
@@ -556,13 +556,12 @@ describe('modules/platform/gitea/index', () => {
         .get(`/repos/${initRepoCfg.repository}`)
         .reply(200, {
           ...mockRepo,
-          allow_rebase: false,
```
Why this change? IMO it makes the test weaker.

dh-seanmurphy replied:
Removed.

gzm0 commented on lib/modules/platform/gitea/index.spec.ts:
FWIW: I agree that the existing tests do not match the existing implementation.

gzm0 reviewed (changes requested):
> Gitea behaviour is to permit merging with an allowed merge style which is not the default merge style; in this case, the merge style is chosen from an ordered list.
> 
> This change permits fallbacks in case the default merge style is not an allowed merge style, falling back to an allowed merge style as per the ordered list. If the default merge style is specified and if it is allowed, it will be selected.

This explanation needs improvement.

> Gitea behaviour is to permit merging with an allowed merge style which is not the default merge style; in this case, the merge style is chosen from an ordered list.

This doesn't explain anything. It says you can merge with an allowed merge style. I'd expect that to be the definition of an allowed merge style.

> This change permits fallbacks in case the default merge style is not an allowed merge style, falling back to an allowed merge style as per the ordered list. If the default merge style is specified and if it is allowed, it will be selected.

There is no explanation why this would be a good thing to do. Notably, there is no mention that this mirrors Gitea's behavior.

dh-seanmurphy commented:
> > Gitea behaviour is to permit merging with an allowed merge style which is not the default merge style; in this case, the merge style is chosen from an ordered list.
> > This change permits fallbacks in case the default merge style is not an allowed merge style, falling back to an allowed merge style as per the ordered list. If the default merge style is specified and if it is allowed, it will be selected.
> 
> This explanation needs improvement.

Thanks - updated - will make modifications to code now and rerequest review.

dh-seanmurphy commented:
@gzm0 - I've significantly refactored this and I've updated the PR desc. The changes since the last review are currently in the top commit; when submitting it, I'd plan on reorganizing commits such that there are just 2 commits, one for code and one for tests.

I've also run this through type checking and verified that the `never` code causes a type checking error if an extra merge method is added to `PRMergeMethod` and the `isAllowed` function is not updated accordingly.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,35 +323,55 @@ const platform: Platform = {
       throw new Error(REPOSITORY_BLOCKED);
     }
 
+    function isAllowed(style: PRMergeMethod, repo: any): boolean {
```
Why take repo as argument? It is in the scope no?

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,35 +323,55 @@ const platform: Platform = {
+    function assertNever(_: never): never {
       logger.debug(
-        'Repository has no allowed merge methods - aborting renovation',
+        'Repo has unknown default merge style - aborting renovation',
```
default is wrong here, this also gets called for hardcoded styles.

dh-seanmurphy replied:
Removed default.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,35 +323,55 @@ const platform: Platform = {
-    ) {
-      config.mergeMethod = 'fast-forward-only';
-    } else {
+    function assertNever(_: never): never {
```
IMO this function name is bordeline in the scope: It is significantly less specific then what it does.

dh-seanmurphy replied:
Changed to `throwUnknownMergeStyle`.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,35 +323,55 @@ const platform: Platform = {
+    }
+
+    const preferredOrder: PRMergeMethod[] = [
+      'rebase',
```
Just add `repo.default_merge_style` as first element of this list?

dh-seanmurphy replied:
Good tip - thanks.

gzm0 commented on lib/modules/platform/gitea/index.ts:
```diff
@@ -323,35 +323,55 @@ const platform: Platform = {
+    ];
+
+    // if default merge style is allowed, use this; else fall back to order
+    // as per preferredOrder
```
This should explain why this is reasonable. Not just what it does.

gzm0 commented on lib/modules/platform/gitea/index.spec.ts:
```diff
@@ -614,29 +594,23 @@ describe('modules/platform/gitea/index', () => {
     });
 
-    it('should fall back to merge method "merge"', async () => {
+    it('should throw if unknown default merge style is configured', async () => {
```
This is a good test case!

dh-seanmurphy commented:
Modified according to feedback - would appreciate another review @gzm0 - thanks.

dh-seanmurphy commented:
> LGTM, I think its time to upstream this.

Great - thanks for v helpful feedback.
